### PR TITLE
Added delegated administratior for Identity Center and AWS Control Tower support feature

### PR DESCRIPTION
**Pull request by ashmeetp**
*Issue #, if available:*

*Description of changes:*

**New feature: Added delegated admin support**
**New feature: Added Control Tower support**

_codepipeline-stack.template:_
- Appended current region and AWS account ID as a suffix to the S3 bucket for automation to keep the bucket name as a unique identifier
- Added permissions to create, update and delete DynamoDB table to CodeBuild IAM role
- Updated CodePipeline pipeline name from _Identity-Center-Automation-Sample-Solution_ to _Identity-Center-Automation-Solution_

_delegate-admin/IC-Delegate-Admin.yml:_
- Added CloudFormation template that creates a custom Lambda function to delegate Organization member account as a delegated administrator for Identity Center

_delegate-admin/IC_delegate_admin_main.py:_
- Added Lambda function that delegates a Organization member account as a delegated administrator for Identity Center. This is the same function that is deployed by the _delegate-admin/IC-Delegate-Admin.yml_ CloudFormation template

_diagram/architecture_diagram.png:_
- Updated the diagram to reflect changes and new components

_identity-center-automation.template:_
- Added additional parameters _OrgManagementAccount_, _AdminDelegated_ and _ControlTowerEnabled_ to add support for delegated administration feature and support Control Tower created permission sets and assignments. Additionally, added conditions to check these parameters, deploy DynamoDB table as necessary and pass the required values as environment variables to Lambda functions _auto-permissionsets_ and _auto-assignment_ 
- Added permissions to add, remove and batch update skipped permission set names and ARNs to the DynamoDB table
- Appended current region and AWS account ID as a suffix to the S3 bucket for automation to keep the bucket name as a unique identifier
- Added _sso:ListAccountsForProvisionedPermissionSet_ permission to Lambda execution IAM role to get the list of accounts for a provisioned Permission Set to support delegated admin feature.
- Added support to create Identity Center administrator IAM role and KMS administrator role as a part of the solution if one doesn't exist in the AWS account already. Added condition to accept either existing roles or the newly created roles to add as an exception to EventBridge rules

_identity-center-s3-bucket.template:_
- Added parameters _createICKMSAdminRole_ and _createICAdminRole_, and a condition to check if Identity Center and KMS administrator role or user exists in teh account, if not, added an option to create as a part of the solution
- Updated S3 bucket and KMS key policy to include the appropriate admin IAM roles as principals
- Exported outputs to be imported by the _identity-center-automation.template_ if new IAM roles are to be created
- Appended current region and AWS account ID as a suffix to the S3 bucket for automation to keep the bucket name as a unique identifier

_identity-center-stacks-parameters.json:_
- Added additional input parameters _AdminDelegated_, _ControlTowerEnabled_, _OrgManagementAccount_, _createICAdminRole_ and _createICKMSAdminRole_ to support delegated administration for Identity Center, Control Tower enabled environments and option to create administrative roles, if those does not exist in the account already

_src/codebuild/buildspec-param.yml:_
- Updated commands to pass the updated parameters to the corresponding CloudFormation stacks 

_src/lambda-code/identity-center-auto-assign/auto-assignment.py:_
- Updated code to skip Permission Sets provisioned in management account if Control Tower is enabled or if a Organization member account is delegated as an administrator for Identity Center. Delegated administrator for Identity Center by design does not have authorization to manage permission sets provisioned in management account. As a workaround, those permission sets are skipped from this automation and the Permission Set names and ARNs are added to a DynamoDB table _IC-SkippedPermissionSets_ for visibility and auditing
- Additionally updated code to skip permission sets tagged with tag key ManagedBy and value ControlTower in order to prevent the automation from removing Permission Sets created by Control Tower. Control Tower requires those Permission Sets to exist in order to create and provision accounts, and if those Permission Sets are removed, will cause issues with new account enrollment to Control Tower. Similar to delegated administrator process, the skipped Permission Sets names and ARNs are added to the DynamoDB table _IC-SkippedPermissionSets_ for visibility and auditing

_src/lambda-code/identity-center-auto-permissionsets/auto-permissionsets.py:_
- Updated code to sync skipped permission sets with DynamoDB table to ensure items in the table are not in drift
- Updated code to skip Permission Sets provisioned in management account if Control Tower is enabled or if a Organization member account is delegated as an administrator for Identity Center. Delegated administrator for Identity Center by design does not have authorization to manage permission sets provisioned in management account. As a workaround, those permission sets are skipped from this automation and the Permission Set names and ARNs are added to a DynamoDB table _IC-SkippedPermissionSets_ for visibility and auditing
- Additionally updated code to skip permission sets tagged with tag key ManagedBy and value ControlTower in order to prevent the automation from removing Permission Sets created by Control Tower. Control Tower requires those Permission Sets to exist in order to create and provision accounts, and if those Permission Sets are removed, will cause issues with new account enrollment to Control Tower. Similar to delegated administrator process, the skipped Permission Sets names and ARNs are added to the DynamoDB table _IC-SkippedPermissionSets_ for visibility and auditing
- Updated code to manage existing Permission Sets without description. The existing Permission Sets that do not have a description will be removed - If solution is deployed in management account, the existing Permission Sets without description will be removed, if solution is deployed in delegated administrator account, the existing Permission Sets without description will be removed as long as they are not provisioned in the management account, else, it will be skipped and will be added to the DynamoDB table for visibility and auditing
- Updated code to remove drift when triggered by EventBridge rule on detecting manual changes to Identity Center

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/lambda-code/identity-center-auto-permissionsets/auto-permissionsets.py [hunk 1]
```diff
@@ -696,86 +863,144 @@ def sync_json_with_aws(local_files, aws_permission_sets, pipeline_id):
...
                 jobId=pipeline_id,
-                failureDetails={'message':str(error), 'type': 'JobFailed'}
+                failureDetails={'message': str(error), 'type': 'JobFailed'}
+            )
+
+    elif event['detail-type'] == 'AWS API Call via CloudTrail':
+        sleep(10)
+        event_detail_type = event['detail-type']
+        try:
+            print("The automation process is now started. This event is triggered by EventBrodge")
```
Comment (fengyueyuzhe) [inline src/lambda-code/identity-center-auto-permissionsets/auto-permissionsets.py:980]: small typo :)

File: delegate-admin/IC-Delegate-Admin.yml [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            Effect: Allow
+            Resource:
+              Fn::GetAtt:
+                - delegateAdminFunctionC209BD88
+                - Arn
+        Version: "2012-10-17"
+      PolicyName: lambdaInvokeCustomResourcePolicyC17933EB
+      Roles:
+        - Ref: AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2
+  AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2:
```
Comment (fengyueyuzhe) [inline delegate-admin/IC-Delegate-Admin.yml:212]: I would recommend to use human-readable resource name

Comment (ashmeetp) [reply delegate-admin/IC-Delegate-Admin.yml:212]: I used CDK to generate that template. I'll change the names. Thanks

File: src/lambda-code/identity-center-auto-permissionsets/auto-permissionsets.py (hunk 1 above)
Comment (ashmeetp) [reply src/lambda-code/identity-center-auto-permissionsets/auto-permissionsets.py:987]: Fixed!

**Review by fengyueyuzhe — APPROVED** [review]
Looks good!

**Pull request merged**